lib/srv/db/clickhouse/test.go

```diff
@@ -0,0 +1,263 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clickhouse
+
+import (
+	"context"
+	"crypto/tls"
+	"database/sql"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httptest"
+
+	"github.com/ClickHouse/ch-go"
+	chproto "github.com/ClickHouse/ch-go/proto"
+	"github.com/ClickHouse/clickhouse-go/v2"
+	"github.com/ClickHouse/clickhouse-go/v2/lib/proto"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+)
+
+// TestServerOption allows setting test server options.
+type TestServerOption func(*TestServer)
+
+// WithClickHouseNativeProtocol specify the Native Protocol for test TestServer.
+func WithClickHouseNativeProtocol() TestServerOption {
+	return func(server *TestServer) {
+		server.protocol = defaults.ProtocolClickHouse
+	}
+}
+
+// WithClickHouseHTTPProtocol specify the HTTP ClickHouse Protocol for test TestServer.
+func WithClickHouseHTTPProtocol() TestServerOption {
+	return func(server *TestServer) {
+		server.protocol = defaults.ProtocolClickHouseHTTP
+	}
+}
+
+// TestServer is a ClickHouse test server that allows to handle
+// basic HTTP and ClickHouse Native connections.
+type TestServer struct {
+	cfg       common.TestServerConfig
+	listener  net.Listener
+	port      string
+	tlsConfig *tls.Config
+	log       logrus.FieldLogger
+	protocol  string
+}
+
+// NewTestServer returns a new instance of a test ClickHouse server.
+func NewTestServer(config common.TestServerConfig, opts ...TestServerOption) (*TestServer, error) {
+	err := config.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tlsConfig, err := common.MakeTestServerTLSConfig(config)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tlsConfig.InsecureSkipVerify = true
+	if config.Listener != nil {
+		config.Listener = tls.NewListener(config.Listener, tlsConfig)
+	}
+	port, err := config.Port()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	testServer := &TestServer{
+		cfg:       config,
+		listener:  config.Listener,
+		port:      port,
+		tlsConfig: tlsConfig,
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: defaults.ProtocolClickHouse,
+			"name":          config.Name,
+		}),
+	}
+
+	for _, opt := range opts {
+		opt(testServer)
+	}
+
+	return testServer, nil
+}
+
+func (s *TestServer) Serve() error {
+	switch s.protocol {
+	case defaults.ProtocolClickHouseHTTP:
+		return trace.Wrap(s.serveHTTP())
+	default:
+		return trace.Wrap(s.serveNative())
+	}
+}
+
+func encodeVersion() ([]byte, error) {
+	var block proto.Block
+	if err := block.AddColumn("versions()", "String"); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := block.Append("23.4.2.11"); err != nil {
```

Comment:
Nit: Any reason for this specific version? Do you know if we have a minimum version Teleport will support? (If so, adding a comment would help here)

Reply:
Added a comment in https://github.com/gravitational/teleport/pull/27004/commits/c2cdb99ffdd8ea6bea3964b0c9880a56cb5d75e8